beacon/beacon.go

```diff
@@ -530,24 +527,29 @@ func (pd *ProtocolDriver) cleanupEpoch(epoch types.EpochID) {
 }
 
 // listens to new layers.
-func (pd *ProtocolDriver) listenLayers(ctx context.Context) {
+func (pd *ProtocolDriver) listenEpochs(ctx context.Context) {
 	pd.logger.With().Info("starting listening layers")
 
+	currentEpoch := pd.clock.GetCurrentLayer().GetEpoch()
+	layer := currentEpoch.Add(1).FirstLayer()
 	for {
 		select {
 		case <-pd.ctx.Done():
 			return
-		case layer := <-pd.layerTicker:
-			pd.logger.With().Debug("received tick", layer)
-			if layer.FirstInEpoch() {
-				epoch := layer.GetEpoch()
-				pd.setProposalTimeForNextEpoch()
-				pd.logger.With().Info("first layer in epoch", layer, epoch)
-				pd.eg.Go(func() error {
-					_ = pd.onNewEpoch(ctx, epoch)
-					return nil
-				})
+		case <-pd.clock.AwaitLayer(layer):
+			current := pd.clock.GetCurrentLayer()
+			if current.Before(layer) {
+				pd.logger.With().Info("time sync detected, realigning Beacon")
+				continue
 			}
```

Comment:
I slightly dislike this check. I feel that it's arbitrary.

In theory, time sync can happen anywhere, so why do we only check for it here? Why not after every row in this function? My point is that just like we're willing to tolerate that this happens right after we check, we should be willing to tolerate that it happens between closing the channel on the timer side and waking up here.

OTOH, there's a different check missing, IMO. If the computer sleeps and wakes up after a while, we might be a long time after the start of the layer (epoch, in this case). I'm guessing there should be a timeout, after which we don't send the proposal because it's too late and everyone else already moved on.

Reply:
I had the following train of thought when I wrote that code:

1. `AwaitLayer` waits until *at least* `layer`, but possibly longer (e.g. hibernation, etc.) so when it returns we have to check in which layer we are actually in.
2. `GetCurrentLayer` returns the current layer which is expected to be the one we waited for, but could be a later one, thats why we use `current` and not `layer` to processing an already passed layer/epoch
3. In *rare* cases `current` is before `layer` (when exactly between `AwaitLayer` and `GetCurrentLayer` the time was adjusted backwards via NTP or by the user). In this case we must prevent beacon to process a layer/epoch that it possibly already handled again.

This `if` should on most systems rarely if ever trigger but if it does it prevents a node from possibly being flagged as malicious.

Reply:
I added a check to verify we are in the first layer of an epoch. If we aren't then the epoch will be skipped by the beacon. This should mimic the behaviour as it was before the changes.